state/kafka-receiver/src/main/java/com/expediagroup/streamplatform/streamregistry/state/kafka/KafkaEventReceiver.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright (C) 2018-2020 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expediagroup.streamplatform.streamregistry.state.kafka;
+
+import static com.expediagroup.streamplatform.streamregistry.state.EventCorrelator.CORRELATION_ID;
+import static com.expediagroup.streamplatform.streamregistry.state.model.event.Event.LOAD_COMPLETE;
+import static io.confluent.kafka.serializers.KafkaAvroDeserializerConfig.SCHEMA_REGISTRY_URL_CONFIG;
+import static io.confluent.kafka.serializers.KafkaAvroDeserializerConfig.SPECIFIC_AVRO_READER_CONFIG;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static lombok.AccessLevel.PACKAGE;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.AUTO_OFFSET_RESET_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.GROUP_ID_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicLong;
+
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+
+import io.confluent.kafka.serializers.KafkaAvroDeserializer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+
+import com.expediagroup.streamplatform.streamregistry.state.EventCorrelator;
+import com.expediagroup.streamplatform.streamregistry.state.EventReceiver;
+import com.expediagroup.streamplatform.streamregistry.state.EventReceiverListener;
+import com.expediagroup.streamplatform.streamregistry.state.avro.AvroConverter;
+import com.expediagroup.streamplatform.streamregistry.state.avro.AvroKey;
+import com.expediagroup.streamplatform.streamregistry.state.avro.AvroValue;
+
+@Slf4j
+@RequiredArgsConstructor(access = PACKAGE)
```

Comment:
Did you consider to use `Kafka Streams`? It has support for views (`KTables`, `GlobalKTables`) and you can expose the stores through an API, correlators (`Joiners`), you wouldn't deal directly with producers and consumers and there is a clear way to deal with more than one partition...

The stream-registry to datahub mappers works in that way. The last version is going to have three stores, this is the design document:

https://pages.github.expedia.biz/eg-data-platform/metadata-ingestion-streams/datahub/design/kafka-streams-modeling/

Reply:
Sorry, I removed to internal link you provided. 

Yes I did. It works for simple use cases, but gets convoluted if your use case doesn't exactly fit the mould. For example we need a `GlobalKTable` so we get the whole dataset in each instance and so it waits for the table to be completely loaded before allowing querying of the KV store. But the topic has to be in the exact KV structure that the KV store needs or you have to build out a transform step into a second topic that the table can source from. I want to keep it simple - 1 topic and avoid coupling to Kstreams impl detail patterns.